src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -515,6 +524,20 @@ private void CompileNamedType(NamedTypeSymbol containingType)
                             {
                                 hasStaticConstructor = true;
                             }
+
+                            if (_moduleBeingBuiltOpt is object && method is SourceMethodSymbolWithAttributes { MethodKind: MethodKind.Ordinary, IsModuleInitializer: true })
+                            {
+                                var rootModuleType = _moduleBeingBuiltOpt.RootModuleType;
+                                rootModuleType.AddModuleInitializerMethod(method);
+
+                                if (rootModuleType.GetSynthesizedMethod(WellKnownMemberNames.StaticConstructorName) is null)
+                                {
+                                    rootModuleType.TryAddSynthesizedMethod(
+                                        new SynthesizedRootModuleTypeStaticConstructor(
+                                            rootModuleType,
+                                            new SynthesizedRootModuleType(_moduleBeingBuiltOpt.SourceModule.GlobalNamespace, rootModuleType)));
+                                }
```

Comment:
Similar to:

https://github.com/dotnet/roslyn/blob/db5c01e59a0fea97281a1adde921c097f48e73e6/src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs#L668-L676

https://github.com/dotnet/roslyn/blob/db5c01e59a0fea97281a1adde921c097f48e73e6/src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs#L1391